File: Training #3 Programacion dinamica/C++/Juanes/collatz.cpp
#include <bits/stdc++.h>

using namespace std;

map<int, int> dict;

int collatz(int n) {
    if (n == 1) return 1;

    if (dict.find(n) != dict.end()) {
        return dict[n];
    }

    int next_n;
    if (n % 2 == 0) {
        next_n = n / 2;
    } else {
        next_n = 3 * n + 1;
    }

    dict[n] = 1 + collatz(next_n);
    return dict[n];
}

int collatz_max_sequence(int n, int m) {
    if (n > m) {
        swap(n, m);
    }

    int max_sequence = 0;
    for (int i = n; i <= m; ++i) {
        int sequence_length = collatz(i);
        if (sequence_length > max_sequence) {
            max_sequence = sequence_length;
        }
    }
    return max_sequence;
}

int main() {
    int n, m;
    while (scanf("%d %d", &n, &m) != EOF) {
        int result = collatz_max_sequence(n, m);
        printf("%d %d %d\n", n, m, result);
    }
    return 0;
}

File: Training #3 Programacion dinamica/Python/Juanes/collatz.py
from sys import stdin

collatz_lengths = {}

def collatz_conjecture(n):
    if n == 1:
        return 1

    if n in collatz_lengths:
        return collatz_lengths[n]

    if n % 2 == 0:
        next_n = n // 2
    else:
        next_n = 3 * n + 1

    collatz_lengths[n] = 1 + collatz_conjecture(next_n)
    return collatz_lengths[n]

def collatz_max_sequence(n, m):
    if n > m:
        n, m = m, n

    max_sequence = 0
    for i in range(n, m + 1):
        sequence_length = collatz_conjecture(i)
        if sequence_length > max_sequence:
            max_sequence = sequence_length

    return max_sequence

def main():
    for line in stdin:
        numbers = line.split()
        n = int(numbers[0])
        m = int(numbers[1])
        result = collatz_max_sequence(n, m)
        print(f"{n} {m} {result}")

main()
